client-tools/Dump/DumpFeature.h

```diff
@@ -135,15 +141,69 @@ class DumpFeature final : public ArangoDumpFeature {
                  std::string const& server,
                  std::shared_ptr<ManagedDirectory::File> file);
 
-    ~DumpShardJob();
-
     Result run(arangodb::httpclient::SimpleHttpClient& client) override;
 
     std::string const shardName;
     std::string const server;
     std::shared_ptr<ManagedDirectory::File> file;
   };
 
+  struct DumpFileProvider {
+    explicit DumpFileProvider(
+        ManagedDirectory& directory,
+        std::map<std::string, arangodb::velocypack::Slice>& collectionInfo,
+        bool splitFiles);
+    std::shared_ptr<ManagedDirectory::File> getFile(
+        std::string const& collection);
+
+   private:
+    struct CollectionFiles {
+      std::size_t count{0};
+      std::shared_ptr<ManagedDirectory::File> file;
+    };
+
+    bool const _splitFiles;
+    std::mutex _mutex;
+    std::unordered_map<std::string, CollectionFiles> _filesByCollection;
+    ManagedDirectory& _directory;
+    std::map<std::string, arangodb::velocypack::Slice>& _collectionInfo;
+  };
+
+  struct ParallelDumpServer : public DumpJob {
+    struct ShardInfo {
+      std::string collectionName;
+    };
+
+    ParallelDumpServer(ManagedDirectory&, DumpFeature&, ClientManager&,
+                       Options const& options, maskings::Maskings* maskings,
+                       Stats& stats, std::shared_ptr<DumpFileProvider>,
+                       std::unordered_map<std::string, ShardInfo> shards,
+                       std::string server);
+
+    Result run(httpclient::SimpleHttpClient&) override;
+
+    std::unique_ptr<httpclient::SimpleHttpResult> receiveNextBatch(
+        httpclient::SimpleHttpClient&, std::uint64_t batchId,
+        std::optional<std::uint64_t> lastBatch);
+
+    void createDumpContext(httpclient::SimpleHttpClient& client);
+    void finishDumpContext(httpclient::SimpleHttpClient& client);
+
+    ClientManager& clientManager;
+    std::shared_ptr<DumpFileProvider> const fileProvider;
+    std::unordered_map<std::string, ShardInfo> const shards;
```

Comment:
Yes reason because is doesn't matter.

Reply:
So instead of option, with same difficulty, smaller allocation count and memory footprint, faster search and insert.
You choose this, because your code spend more time in the IO.

Do I understand you right?

Reply:
Sadly you don't.

Reply:
Ok, explain please